File: array_product_except_self.cpp
// Time Complexity : O(n)
// Space Complexity :  O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this - 

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        
        int i;
        vector<int> ret_vect;
        int len = nums.size();
        
        // Variable for running product
        int p = 1;

        // Traverse in Forward Direction        
        for(i = 0; i < len; i++)
        {
            ret_vect.push_back(p);
            p *= nums[i];
        }
        
        p = 1;
        
        // Traverse in Reverse Direction
        for(i = len-1; i >= 0; i--)
        {
            ret_vect[i] *= p;
            p *= nums[i];
        }
        
        return ret_vect;
    }
};

File: Array_ProductExceptSelf.cs
using System;
namespace Algorithms
{
    /// Time Complexity : O(N) 
    // Space Complexity :O(1) 
    // Did this code successfully run on Leetcode :Yes
    // Any problem you faced while coding this :  No
    public class Array_ProductExceptSelf
    {
        public int[] ProductExceptSelf(int[] nums)
        {
            int length = nums.Length;
            int[] result = new int[length];


            result[0] = 1;

            for (int i = 1; i < length; i++)
            {
                result[i] = result[i - 1] * nums[i - 1];
            }
            int right = 1;

            for (int i = length - 1; i >= 0; i--)
            {
                result[i] = result[i] * right;
                right *= nums[i];
            }

            return result;
        }
    }
}
